### PR TITLE
add C FFI

### PR DESCRIPTION
**Pull request by Lupino**

### CODE REVIEW

File: src/Data/Unfoldable1.c [hunk 1]
```diff
@@ -0,0 +1,15 @@
+#include <purescript.h>
+
+PURS_FFI_FUNC_6(Data_Unfoldable1_unfoldr1ArrayImpl, isNothing, fromJust, fst, snd, f, b, {
+	purs_vec_t * result = (purs_vec_t *) purs_vec_new();
+	purs_any_t * value = (purs_any_t *) b;
```
Comment (felixSchl) [inline src/Data/Unfoldable1.c]: no need for this line, just name `b` in the arguments "value". we're not mutating it.

File: src/Data/Unfoldable1.c [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+	while (1) {
+		const purs_any_t * tuple = purs_any_app(f, value);
+		purs_vec_push_mut(result, purs_any_app(fst, tuple));
+		const purs_any_t * maybe = purs_any_app(snd, tuple);
+		if (purs_any_is_true(purs_any_app(isNothing, maybe)) == purs_any_true) break;
```
Comment (felixSchl) [inline src/Data/Unfoldable1.c]: while will need updating to:

```
		if (purs_any_is_true(purs_any_app(isNothing, maybe))) {
			break;
		}
```

File: src/Data/Unfoldable.c [hunk 3]
```diff
@@ -0,0 +1,14 @@
+#include <purescript.h>
+
+PURS_FFI_FUNC_6(Data_Unfoldable_unfoldrArrayImpl, isNothing, fromJust, fst, snd, f, b, {
+	purs_vec_t * result = (purs_vec_t *) purs_vec_new();
+	purs_any_t * value = (purs_any_t *) b;
+	while (1) {
+		const purs_any_t * maybe = purs_any_app(f, value);
+		if (purs_any_is_true(purs_any_app(isNothing, maybe)) == purs_any_true) break;
```
Comment (felixSchl) [inline src/Data/Unfoldable.c]: if (purs_any_is_true(purs_any_app(isNothing, maybe))) {
			break;
		}

**Comment by felixSchl** [comment]
Let's also add a way to run the tests. This package shows how: https://github.com/pure-c/purescript-foldable-traversable. Essentially adding a Makefile which expects `PUREC_DIR` be set (it will error out otherwise). Then you can run the tests using `make -j16 && ./main.out`, or we could add phony a `make check` target that does the same

**Comment by Lupino** [comment]
@felixSchl if we need to update bower.json for the test?

**Comment by felixSchl** [comment]
It would, of course, be preferable if there was a way to piggy back off the original repos long term, thereby outsourcing maintenance of those libraries (except C FFI bindings) to the original maintainers. But maybe that's something that can be addressed with tooling? How can we bower install from a fork and still keep using tags for versions? It would be nice we could keep the same tags as the original repos, except we'd update the FFI bindings. So say purescript-xyz releases a new tag 1.1, and we're currently at 1.0 of that library, we'd rebase onto 1.1 and update the bindings, then release a corresponding 1.1 of our fork. I hope we'll be able to add tooling to make this job bearable.

**Comment by felixSchl** [comment]
I think we could perhaps set the "name" field of our forks' bower.json to "purec-&lt;name>", register them on bower, and then set our deps names to e.g.: `{ ... "purec-prelude": "1.1"`?

**Pull request merged**